Pull request for supervisor review (#1)
AstroViking/ba-thesis: chapters/03-implementation.Rmd

>     components_distribution=f(data))

Reviewer: [6] Line 5:  simple -> exact?

```

> Interestingly, due to the smart encapsulated structure of TensorFlow Probability we can use any distribution of the loc-scale family type as a kernel, if there exists an implementation for it in TensorFlow Probability. If the used Kernel has only bounded support, the implementation proposed in this paper allows to specify the support upon instantiation of the class. If the Kernel has infinite support (like a Gaussian kernel for instance) a practical support estimate is calculated by searching for approximate roots with Brent's method[@brent1971algorithm] implemented for TensorFlow in the python package `tf_quant_finance` by Google. This allows us to speed up the calculation.

Reviewer: ```suggestion
Interestingly, due to the smart encapsulated structure of TensorFlow Probability we can use any distribution of the loc-scale family type as a kernel as long as it follows the Distribution contract in TensorFlow Probability. If the used Kernel has only bounded support, the implementation proposed in this paper allows to specify the support upon instantiation of the class. If the Kernel has infinite support (like a Gaussian kernel for instance) a practical support estimate is calculated by searching for approximate roots with Brent's method[@brent1971algorithm] implemented for TensorFlow in the python package `tf_quant_finance` by Google. This allows us to speed up the calculation.
```

> However calculating an exact kernel density estimation is not always feasible as this can take a long time with a huge collection of events, especially in high energy physics. By implementing it in TensorFlow we already get a significant speed up compared to implementations in native Python, since most of TensorFlow is actually implemented in C++ and the code is optimized before running. But the computational complexity however, remains the same of course.

Reviewer: [7] Try to be more specific on the terminology: instead of events, what's the exact term? Datapoints? It doesn't matter, be consistent (but "events" is HEP specific, right?)

> As described in the excellent overview by Artur Gramacki[@gramacki2018fft] simple binning or linear binning can be used, although the last is often preferred since it is more accurate and the difference in computational complexity is negligible.

Reviewer: ```suggestion
As described in the extensive overview by Artur Gramacki[@gramacki2018fft] simple binning or linear binning can be used, although the last is often preferred since it is more accurate and the difference in computational complexity is negligible.
```

Don't judge ;) "excellent" is quite subjective. Not saying that it is not possible to write _if it is truly excellent_ **objectively**. But is it?

(\#eq:linbin)
\end{equation}

> Implementing linear binning efficiently with TensorFlow is a bit tricky since loops should be avoided. However with some inspiration from the excellent KDEpy package[@KDEpy] this can be done without using loops at all. By transforming the data such that every data point $x_k$ can be described by an integral part (corresponding to its nearest left grid point number $l$) plus some fractional part (corresponding to the distance between grid point $g_l$ and data point $x_k$) and applying `tf.math.bincount` twice to the transformed integral part weighting it with the fractional part times the initial weight.

Reviewer: [5] This is rather on the too specific side, especially when the method is mentioned. It should be understandable for the reader without having to look at the TensorFlow implementation

> 

Reviewer: [9] Line 94: yeah, that maybe true that it's difficult (for you and me), but it doesn't matter (don't judge). Does the book really add something? Mentioning it?

Reviewer: Since it is indeed difficult, can you give a qualitative overview of the _idea_? It gets quite mathematical afterwards

(\#eq:tstar)
\end{equation}

Using the fact that $\|f^{(j)}\|^2 = (-1)^j \mathbb{E}_f[f^{(2j)}(X)], \text{ } j\geq 1$ and an initial estimation for $\|\widehat{f}_{h,norm}^{(l+2)}\|^2$  for some $l \geq 3$ one can then iteratively get a an estimation for $\|\widehat{f}_{h,norm}^{(2)}\|^2$ which can then be used to estimate $t^{\ast}$ instead of $\|f''\|^2$. According to their handbook $l = 7$ is a suitable value to yield good practical results.

> The improvment compared to the standard Sheather-Jones plug-on method[@sheather1991reliable] consists in the fact that the initial estimation of $\|\widehat{f}_{h,norm}^{(l+2)}\|^2$ is calculated by solving the partial differental equation using the fast cosine transform as described above, eliminating the need to assume normally distributed data for the initial estimate and leading to improved performance, especially for density distributions that are far from normal as seen in the next chapter.

Reviewer: [6] you did not mention the standard one before? Don't you want to first say something about it before you introduce the improved version? E.g. explain the default and then explain the improved? This would seem more natural to me

> The implementation of the algorithm in TensorFlow proposed in this paper was also inspired a lot by the python package KDEpy[@KDEpy] and uses Brent's method[@brent1971algorithm] to find roots, implemented in TensorFlow in the python package `tf_quant_finance`.

Reviewer: I don't quite see the meaning of this sentence, it seems somewhat out of context?

> One shortcoming of the improved Sheather-Jones algorithm (ISJ) is that with few data points to estimate from, it is not guarenteed to converge. If that happens, one has to use the exact, binned or FFT kernel density estimators as described above.

Reviewer: [7] what are "few" roughly?
